### PR TITLE
Space Storage

### PR DESCRIPTION
**Pull request by mattnibs**
Make package Storage, responsible for managing all log data for a space.

Eventually there should be a central interface for Storage so that
a space can support multiple storage backends. For now this is limited
to a single ZngFile storage backend where:
- Data is populated by passing a zbuf.Reader into it.
- Data can only be rewritten. It cannot be updated.
- Data is housed in a single all.zng file.
- Concurrent reads are allowed but only one active write is allowed.
- Maintains an active cache of the total time span of all records stored.

Additional supporting changes:
- Reworked ingest.Log to instead return a LogIngestOp struct that exports
go channels for "Warnings" and "Done" notifications. Also includes
a Stats() function that provides a thread-safe snapshot of current ingest
stats.
- Reworked ingest.Pcap to use space.Storage for generating log snapshots.

### CODE REVIEW

File: zio/zngio/sort.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
+package zngio
```
Comment (alfred-landrum) [inline zio/zngio/sort.go:1]: Like we discussed in slack, this sort implementation will get removed, and would need to be in a separate PR where it would be considered to replace the existing logic in the sort proc.

File: zbuf/batch.go [hunk 2]
```diff
@@ -48,3 +48,52 @@ func ReadBatch(zr Reader, n int) (Batch, error) {
 	}
 	return NewArray(recs, nano.NewSpanTs(minTs, maxTs)), nil
 }
+
+func ReadBatchSize(zr Reader, n int64) (Batch, error) {
```
Comment (alfred-landrum) [inline zbuf/batch.go:52]: This can be removed with the sort code.

File: zio/detector/reader.go [hunk 3]
```diff
@@ -59,6 +60,52 @@ func NewReader(r io.Reader, zctx *resolver.Context) (zbuf.Reader, error) {
 	return nil, joinErrs([]error{tzngErr, zeekErr, ndjsonErr, zjsonErr, zngErr})
 }
 
+func Format(r io.Reader) (string, error) {
```
Comment (alfred-landrum) [inline zio/detector/reader.go:63]: I don't see why you've duplicated all of the detection logic. This is only called from logOp.openFile; there should be a way for it to call detector.OpenFile.

File: zbuf/zng.go [hunk 4]
```diff
@@ -97,3 +97,70 @@ func CopyWithContext(ctx context.Context, dst WriteFlusher, src Reader) error {
...
+	return nil, nil
+}
+
+func MultiReader(readers ...Reader) Reader {
+	r := make([]Reader, len(readers))
+	copy(r, readers)
+	return &multiReader{readers: r}
+}
+
+type namedReader struct {
```
Comment (alfred-landrum) [inline zbuf/zng.go:155]: I don't think this is worth putting in the zbuf package. It's called once from LogOp.openFile. I don't understand how it can be used since NamedReader returns a Reader interface which doesn't expose String()?

File: zqd/ingest/log.go [hunk 5]
```diff
@@ -5,207 +5,138 @@ import (
...
-	}
-	return pipe.Send(&api.LogPostStatus{
-		Type:         "LogPostStatus",
-		LogTotalSize: d.totalSize,
-		LogReadSize:  logBytesRead(),
-	})
+	// Wrap zbuf.Reader in NamedReader (to make file easier to
+	// identify for errors) and WarningReader (so Read warnings propogate to
+	// warning channel).
+	zr = zbuf.NamedReader(zr, path)
```
Comment (alfred-landrum) [inline zqd/ingest/log.go:97]: I don't see how NamedReader's String() method could be called.

File: zqd/ingest/log.go [hunk 6]
```diff
@@ -5,207 +5,138 @@ import (
...
-			n += readCounters[i].nread
-		}
-		return n
-	}
+func (p *LogOp) openWarning(path string, err error) {
+	p.warnings = append(p.warnings, fmt.Sprintf("%s: %s", path, err))
+}
 
-	zngfile, err := s.CreateFile(filepath.Join(tmpIngestDir, allZngTmpFile))
+func (p *LogOp) openFile(zctx *resolver.Context, path string, cfg detector.OpenConfig) error {
```
Comment (alfred-landrum) [inline zqd/ingest/log.go:62]: This function does too much, and all of the reader/writer plumbing that's going on is hard to follow. I think the former setup was easier to follow: loop over req.path, and do 3 separate things: open the file, calling the detector package to get a zbuf.Reader from it, then add the reader & info to the readers state.

File: zqd/storage/zng.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+	if first {
+		w.span = s
+	} else {
+		w.span = w.span.Union(s)
+	}
+	return nil
+}
+
+// XXX only one writeto allowed at a time
+func (s *ZngStorage) WriteTo(ctx context.Context, zr zbuf.Reader) error {
```
Comment (alfred-landrum) [inline zqd/storage/zng.go:84]: As I suggested previously, renaming this to "Replace" would better communicate what it currently does.

File: zqd/space/space.go [hunk 8]
```diff
@@ -44,26 +40,25 @@ func newSpaceID() api.SpaceID {
 }
 
 type Space struct {
+	Storage *storage.ZngStorage
+
 	path string
 	conf config
-
-	index *zngio.TimeIndex
+	mu   sync.RWMutex
```
Comment (alfred-landrum) [inline zqd/space/space.go:47]: Please remove the opMutex, deletePending, and other associated changes, if needed, those can happen in a separate PR.

File: zio/detector/reader.go (hunk 3 above)
Comment (mattnibs) [reply zio/detector/reader.go:63]: This is the essence of detector should be about. Aside from this its a little complicated why I did this: In log ingest endopoint, if a file was getting rejected by detector.NewReader, the bytes read would get counted towards the total of bytes read in the log ingest op. To me the simplest solution was to use this function to detetect the stream type then set the underlying file seek to `SeekStart` and pass the reset file into `LookupReader`.

File: zqd/ingest/log.go (hunk 5 above)
Comment (mattnibs) [reply zqd/ingest/log.go:97]: If it's the last reader wrapped then it works. Warning reader should happen before this.

File: zqd/storage/zng.go (hunk 7 above)
Comment (mattnibs) [reply zqd/storage/zng.go:84]: Yep.

File: zqd/space/space.go (hunk 8 above)
Comment (mattnibs) [reply zqd/space/space.go:47]: Sure.

File: zqd/ingest/log.go (hunk 6 above)
Comment (mattnibs) [reply zqd/ingest/log.go:62]: I agree that this function could be broken up. OTH IMHO the way it was was harder to follow. Instead you had the `ingestLogs` function that was doing too much.

File: zqd/ingest/log.go (hunk 5 above)
Comment (mattnibs) [reply zqd/ingest/log.go:97]: I was trying to model the wonder flexibility of io package- the ability to wrap and compose interfaces.

File: zqd/space/space.go (hunk 8 above)
Comment (mattnibs) [reply zqd/space/space.go:47]: Not needed, but simpler with less moving parts.

File: zio/detector/reader.go (hunk 3 above)
Comment (mattnibs) [reply zio/detector/reader.go]: Dropped it and went back to readCounter.

File: zqd/ingest/log.go (hunk 5 above)
Comment (mattnibs) [reply zqd/ingest/log.go]: Got rid of it.

File: zqd/ingest/log.go (hunk 6 above)
Comment (mattnibs) [reply zqd/ingest/log.go]: Reverted.

File: zbuf/zng.go (hunk 4 above)
Comment (mattnibs) [reply zbuf/zng.go]: Removed.

**Comment by mattnibs** [comment]
The operations on ZngStorage.Span() are all not thread safe and they should be. I do not think they were beforehand. In the interest of blocking other work I thought I'd put this up, but am willing to fix that here if people think I should.

File: zqd/ingest/log_test.go [hunk 9]
```diff
@@ -33,6 +34,7 @@ func writeTempFile(t *testing.T, data string) string {
 }
 
 func TestLogsErrInFlight(t *testing.T) {
+	t.Skip("Flaky, racy tests. Redo this.")
```
Comment (mattnibs) [inline zqd/ingest/log_test.go:37]: While this was passing every time on my macbook pro, on github actions this reliably fails. This used to normally pass but in this change we don't know about concurrent writes until inget.LogIngestOp calls storage.Replace.

Writing a reliable test case for this is going to require more plumbing.

File: zqd/api/api.go [hunk 10]
```diff
@@ -109,7 +109,6 @@ type PcapPostStatus struct {
 
 type LogPostRequest struct {
 	Paths          []string             `json:"paths"`
-	StopErr        bool                 `json:"stop_err"`
```
Comment (alfred-landrum) [inline zqd/api/api.go:112]: Removing this option from the api can happen in separate PR.

File: zqd/ingest/log.go [hunk 11]
```diff
@@ -3,79 +3,86 @@ package ingest
...
 // Like ingest.Pcap, this overwrites any existing data in the space.
-func Logs(ctx context.Context, pipe *api.JSONPipe, s *space.Space, req api.LogPostRequest) error {
-	ingestDir := s.DataPath(tmpIngestDir)
-	if err := os.Mkdir(ingestDir, 0700); err != nil {
-		// could be in use by pcap or log ingest
-		if os.IsExist(err) {
-			return ErrIngestProcessInFlight
-		}
-		return err
+func NewLogOp(ctx context.Context, store *storage.ZngStorage, req api.LogPostRequest) (*LogOp, error) {
```
Comment (alfred-landrum) [inline zqd/ingest/log.go:35]: Interactions with LogOp have 3 different ways to give feedback/warnings/errors:
- an error from NewLogOp
- warnings sent back via channel
- final error sent when done channel closed

Since all the warnings are generated at setup time in NewLogOp, consider changing NewLogOp to:
```go
func NewLogOp(ctx context.Context, store *storage.ZngStorage, req api.LogPostRequest) (op *LogOp, warnings []error, err error)
```
and send all the warnings immediately after the call; that will remove the warnings channel & associated code.

You can also change the done channel to a `chan error`, so the select on it can be:
```go
case err := <-Done()
```

File: zqd/storage/zng.go [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+	}
+	defer s.wsem.Release(1)
+	if err := os.Remove(s.join(allZngFile)); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	return s.UnsetSpan()
+}
+
+// XXX This is not thread safe and it should be.
+func (s *ZngStorage) SetSpan(span nano.Span) error {
```
Comment (alfred-landrum) [inline zqd/storage/zng.go:163]: The behavior for this function doesn't match the name, and in pcap ingest, and this behavior requires the `UnsetSpan` , `SetSpan` sequence during pcap ingest. Can the unioning be removed so this function just sets the space's span?

File: zqd/storage/zng.go [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+		f, err = os.Open(s.join(bzngFile))
+		if err != nil {
+			if !os.IsNotExist(err) {
+				return nil, err
+			}
+			r := zngio.NewReader(strings.NewReader(""), zctx)
+			return zbuf.NopReadCloser(r), nil
+		}
+	}
+	// XXX plz keep track of how long this takes
```
Comment (alfred-landrum) [inline zqd/storage/zng.go:80]: You can remove this comment.

File: zqd/api/api.go (hunk 10 above)
Comment (mattnibs) [reply zqd/api/api.go:112]: ok

File: zqd/ingest/log.go (hunk 11 above)
Comment (mattnibs) [reply zqd/ingest/log.go:35]: @alfred-landrum warnings can also be generated if a reader comes across an invalid record while reading, or runs into a spontaneous error: see line 58

File: zqd/storage/zng.go (hunk 12 above)
Comment (mattnibs) [reply zqd/storage/zng.go:163]: The unioning is needed because otherwise spans from the pcap ingests snapshots would overwrite the the full span size we get from generating the pcap index. I think the best long term solution would be to have space.Span() take the union of PcapStorage.Span and LogStorage.Span, but I put this in for now to arrive at feature parity. I should add a comment mentioning this here.

File: zqd/ingest/log_test.go [hunk 14]
```diff
@@ -42,30 +43,18 @@ func TestLogsErrInFlight(t *testing.T) {
...
-		return
-	}
-	assert.Fail(t, "expected only one error")
+	op1, err1 := ingest.NewLogOp(ctx, s.Storage, api.LogPostRequest{Paths: []string{f}})
+	op2, err2 := ingest.NewLogOp(ctx, s.Storage, api.LogPostRequest{Paths: []string{f}})
+	require.NoError(t, err1)
+	require.NoError(t, err2)
+	assert.NoError(t, op1.Error())
+	require.Error(t, op2.Error())
+	assert.Contains(t, op2.Error().Error(), storage.ErrWriteInProgress.Error())
```
[the remark this answers is not in the document]
Comment (mattnibs) [reply zqd/ingest/log_test.go:60]: LogOp doesn't close all goroutines when done- I don't believe this is necessary either.

File: zqd/ingest/log.go (hunk 11 above)
Comment (mattnibs) [reply zqd/ingest/log.go:35]: w.r.t making Done and `chan error`, the down side to this approach is that only one subscriber will know if an error occurred. If I make the channel a buffered channel there won't even need to be a subscriber in order for the error to be received by consumers.

I suspect this is why the authors of package context took this approach.

File: zqd/ingest/log_test.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (mattnibs) [reply zqd/ingest/log_test.go]: That said perhaps closing the goroutines is the most hygienic thing to do (I'm a little fuzzy on the idiomatic go approach here)? Either way I've done away with this test and put a similar test in zqd/storage- can add a test for your case if you want.

File: zqd/ingest/log.go (hunk 11 above)
Comment (alfred-landrum) [reply zqd/ingest/log.go:35]: ok, how about removing the error return from NewLogOp entirely, and having it use a single channel to return warnings on, that when closed, signals the operation is done? So inside log:

```go
  op := NewLogOp(ctx, store, req)
  ...
  select {
  case warning, ok := <- op.Status()
  if !ok {
     // op has finished, get final error if any
     err := op.Error()
     ...
     return
  }
  // Otherwise, send warning:
  err := pipe.Send(api.LogPostWarning{...})
  ...
  }
```

That would reduce the op to use a single channel, and remove the extra code to handle storing & draining the warnings.

File: zqd/storage/zng_test.go [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+)
+
+type waitReader time.Duration
+
+func (w waitReader) Read() (*zng.Record, error) {
+	time.Sleep(time.Duration(w))
+	return nil, errors.New("time out")
+}
+
+func TestFailOnConcurrentWrites(t *testing.T) {
```
Comment (alfred-landrum) [inline zqd/storage/zng_test.go:23]: 👍

File: zqd/storage/zng.go (hunk 12 above)
Comment (alfred-landrum) [reply zqd/storage/zng.go:162]: Then for now, I suggest you change the name from `SetSpan` to `UnionSpan`, as the behavior is surprising given the name.

File: zqd/ingest/log.go (hunk 11 above)
Comment (mattnibs) [reply zqd/ingest/log.go:35]: Yep, I really like that idea.

File: zqd/storage/zng.go (hunk 12 above)
Comment (mattnibs) [reply zqd/storage/zng.go:162]: Agreed, much clearer.

**Comment by mattnibs** [comment]
> Would you please create a draft Brim PR using this branch's head, so that we can verify that the Brim integration tests pass with this change?

@alfred-landrum https://github.com/brimsec/brim/pull/772
